app/src/main/cpp/WidgetMover.cpp

```diff
@@ -0,0 +1,151 @@
+/* -*- Mode: C++; tab-width: 20; indent-tabs-mode: nil; c-basic-offset: 2 -*-
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "WidgetMover.h"
+#include "Widget.h"
+#include "WidgetPlacement.h"
+#include "VRBrowser.h"
+#include "vrb/ConcreteClass.h"
+#include "vrb/Matrix.h"
+
+namespace crow {
+
+// Should match the values defined in WidgetManagerDelegate.WidgetMoveBehaviourFlags
+enum class WidgetMoveBehaviour {
+  GENERAL = 0,
+  KEYBOARD = 1
+};
+
+struct WidgetMover::State {
+  WidgetPtr widget;
+  int attachedController;
+  vrb::Vector initialPoint;
+  vrb::Matrix initialTransform;
+  WidgetPlacementPtr initialPlacement;
+  WidgetPlacementPtr movePlacement;
+  WidgetMoveBehaviour moveBehaviour;
+  vrb::Vector endDelta;
+  float endRotation;
+
+  State()
+      : widget(nullptr)
+      , attachedController(-1)
+      , moveBehaviour(WidgetMoveBehaviour::GENERAL)
+      , endRotation(0)
+  {}
+
+
+  WidgetPlacementPtr HandleKeyboardMove(const vrb::Vector& aDelta) {
```

Comment:
nit: you can probably return `WidgetPlacementPtr&` to reduce incrementing/decrementing the `std::shared_ptr<>`.